File: makehost.py
import os
import textwrap
import datetime

nl = str("\n")
space = str(" ")
line_break = str("#=========\n")


def makehost(
        host_name: str,
        host_title: str,
        host_desc: str,
        domains_with_www_prefix: set,
        domains_sans_www_prefix: set,
        directory_output: str = ""
        ):
    # Sort the lists out alphanumerically
    #
    domains_sans_www_prefix = sorted(domains_sans_www_prefix)
    domains_with_www_prefix = sorted(domains_with_www_prefix)
    # Info for each host file
    #
    host_comment = textwrap.dedent("""\
    # Title:
    # *host_title*
    #
    # Description:
    # *host_desc* (simple as that!)                    
    #                         
    # Compatible with AdAway on Android and multiple ad blockers.
    #
    # Source(s) Used:
    # https://raw.githubusercontent.com/ryanbarillosofficial/dns-blocklists/main/search_engines/sources.json
    #
    # Project Home Page:
    # https://github.com/ryanbarillos/dns-blocklists
    #
    """)

    # Update the "host_comment" for appropriate host file
    #
    comment = host_comment.replace("*host_title*", host_title).replace("*host_desc*", host_desc)

    # Delete the host file if exists already for replacing
    #
    file_name: str = host_name + ".txt"
    if os.path.exists(file_name): os.remove(file_name)

    # Make a new host file of the same name
    #
    file = open(file_name, "x")
    file.write(comment)
    file.write("# Last Updated: " + datetime.date.today().strftime("%d %b %Y") + nl) 
    # IPv4 addresses
    #
    file.write("\n\n" + line_break + "IPv4 Nodes\n" + line_break)    
    for domain in domains_sans_www_prefix:
        file.write("0.0.0.0" + space + domain + nl)
    for domain in domains_with_www_prefix:
        file.write("0.0.0.0" + space + domain + nl)
    # IPv6 addresses
    #
    file.write("\n\n" + line_break + "IPv6 Nodes\n" + line_break)    
    for domain in domains_sans_www_prefix:
        file.write("::1" + space + domain + nl)
    for domain in domains_with_www_prefix:
        file.write("::1" + space + domain + nl)

File: makehost.js
// Imports
import { exit } from "process";
import fs from "fs";

// Constant variables
const DIRECTORY_CURRENT = process.cwd();
const BREAK_BLOCK = "\n\n\n";
const BREAK_LINE = "#===============";
const IP_PREFIX = {
  V4: "0.0.0.0",
  V6: "::1",
};

export default function makehost(
  host_name = "all",
  host_title,
  host_description,
  domains_sans_www_prefix,
  domains_with_www_prefix,
  directory_output = ""
) {
  /*
  Convert sets into arrays for iteration
  And then sort them alphabetically
  */
  const DOMAINS_WITH_WWW_PREFIX = Array.from(domains_with_www_prefix).sort(
    (a, b) => a - b
  );
  const DOMAINS_SANS_WWW_PREFIX = Array.from(domains_sans_www_prefix).sort(
    (a, b) => a - b
  );

  // Info for each host file
  const FILE_NAME = `${host_name}.txt`;
  const FILE_PATH = `${DIRECTORY_CURRENT}/${directory_output}/${FILE_NAME}`;
  const HOST_COMMENT = `# Title: \
    \n# ${host_title} \
    \n# \
    \n# Description: \
    \n# ${host_description} — simple as that! \
    \n# \
    \n# Compatible with AdAway on Android and multiple ad blockers. \
    \n# \
    \n# Source(s) Used: \
    \n# https://raw.githubusercontent.com/ryanbarillosofficial/hosts/main/${directory_output}/sources.json \
    \n# \
    \n# Number of Unique Domains (without their "www." variants): \
    \n# ${domains_sans_www_prefix.size} \
    \n# \
    \n# Project Home Page: \
    \n# https://github.com/ryanbarillos/hosts \
    \n#${BREAK_BLOCK}`;

  // String of each domain blocked
  const IPV4_ADDRESSES =
    `${BREAK_LINE}\n# IPv4 Addresses\n${BREAK_LINE}\n` +
    (() => {
      let output = "";
      for (let i = 0; i < DOMAINS_WITH_WWW_PREFIX.length; i++) {
        output += `${IP_PREFIX.V4} ${DOMAINS_SANS_WWW_PREFIX[i]}\n`;
        output += `${IP_PREFIX.V4} ${DOMAINS_WITH_WWW_PREFIX[i]}\n`;
      }
      return output;
    })() +
    BREAK_BLOCK;
  const IPV6_ADDRESSES =
    `${BREAK_LINE}\n# IPv6 Addresses\n${BREAK_LINE}\n` +
    (() => {
      let output = "";
      for (let i = 0; i < DOMAINS_WITH_WWW_PREFIX.length; i++) {
        output += `${IP_PREFIX.V6} ${DOMAINS_SANS_WWW_PREFIX[i]}\n`;
        output += `${IP_PREFIX.V6} ${DOMAINS_WITH_WWW_PREFIX[i]}\n`;
      }
      return output;
    })();

  // Make the host text file
  fs.writeFile(
    FILE_PATH,
    HOST_COMMENT + IPV4_ADDRESSES + IPV6_ADDRESSES,
    (err) => {
      if (err) {
        console.error(err);
      } else {
        console.log(`Created \"${FILE_NAME}\" for \"${host_title}\" — yay!`);
      }
    }
  );
}
